<a href="https://colab.research.google.com/github/reesha-rsh/MLb4/blob/main/Homework/Homework_2_0_Pandas_Toxic_Comments_EDA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [40]:
#basics
import pandas as pd
import numpy as np

In [41]:
import datetime

In [122]:
import matplotlib
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline
plt.style.use('fivethirtyeight')

In [43]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [98]:
train = pd.read_csv("/content/drive/MyDrive/MLb4/Homework/2.0 Pandas - Toxic Comments EDA/train_1.csv")

In [99]:
train.head()

Page  2015-07-01  2015-07-02  \
0            2NE1_zh.wikipedia.org_all-access_spider        18.0        11.0   
1             2PM_zh.wikipedia.org_all-access_spider        11.0        14.0   
2              3C_zh.wikipedia.org_all-access_spider         1.0         0.0   
3         4minute_zh.wikipedia.org_all-access_spider        35.0        13.0   
4  52_Hz_I_Love_You_zh.wikipedia.org_all-access_s...         NaN         NaN   

   2015-07-03  2015-07-04  2015-07-05  2015-07-06  2015-07-07  2015-07-08  \
0         5.0        13.0        14.0         9.0         9.0        22.0   
1        15.0        18.0        11.0        13.0        22.0        11.0   
2         1.0         1.0         0.0         4.0         0.0         3.0   
3        10.0        94.0         4.0        26.0        14.0         9.0   
4         NaN         NaN         NaN         NaN         NaN         NaN   

   2015-07-09  ...  2016-12-22  2016-12-23  2016-12-24  2016-12-25  \
0        26.0  ...        32.0        63.0        15.0        26.0   
1        10.0  ...        17.0        42.0        28.0        15.0   
2         4.0  ...         3.0         1.0         1.0         7.0   
3        11.0  ...        32.0        10.0        26.0        27.0   
4         NaN  ...        48.0         9.0        25.0        13.0   

   2016-12-26  2016-12-27  2016-12-28  2016-12-29  2016-12-30  2016-12-31  
0        14.0        20.0        22.0        19.0        18.0        20.0  
1         9.0        30.0        52.0        45.0        26.0        20.0  
2         4.0         4.0         6.0         3.0         4.0        17.0  
3        16.0        11.0        17.0        19.0        10.0        11.0  
4         3.0        11.0        27.0        13.0        36.0        10.0  

[5 rows x 551 columns]

In [46]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 145063 entries, 0 to 145062
Columns: 551 entries, Page to 2016-12-31
dtypes: float64(550), object(1)
memory usage: 609.8+ MB


In [47]:
import re
import string

In [48]:
#train["PageName"] = train["Page"].str.split(".", expand = True)[0]

In [49]:
#train["Language"] = train["PageName"].apply(lambda x: (re.split("_",str(x)))[-1])

In [50]:
#train["Language"]  = train["Page"].apply(lambda x: (re.findall("_\w{2}\.",str(x))))

In [51]:
#train[(train["Page"].str.contains("zh")==True) & (train["Page"].str.contains("wikipedia")==False)]

In [100]:
train["Language"] = train["Page"].apply(lambda x: (re.search("_(\w{2})\.wikipedia\.org",str(x)))).apply(lambda x: str(x.group(1)) if x else None )

In [53]:
train[train["Language"].isnull()]


Page  2015-07-01  \
13332    Accueil_commons.wikimedia.org_all-access_spider        55.0   
13333  Atlas_of_Asia_commons.wikimedia.org_all-access...         5.0   
13334  Atlas_of_Europe_commons.wikimedia.org_all-acce...         4.0   
13335  Atlas_of_World_War_II_commons.wikimedia.org_al...         5.0   
13336  Atlas_of_colonialism_commons.wikimedia.org_all...         8.0   
...                                                  ...         ...   
84730  help:contents_www.mediawiki.org_all-access_spider         NaN   
84731  how_to_contribute_www.mediawiki.org_all-access...         0.0   
84732      mediawiki_www.mediawiki.org_all-access_spider         0.0   
84733   special:book_www.mediawiki.org_all-access_spider         0.0   
84734  special:translate_www.mediawiki.org_all-access...         0.0   

       2015-07-02  2015-07-03  2015-07-04  2015-07-05  2015-07-06  2015-07-07  \
13332        48.0        44.0        35.0        46.0        41.0        32.0   
13333         3.0         4.0         6.0         3.0         1.0         5.0   
13334         6.0         9.0         7.0         6.0         4.0         5.0   
13335         2.0         6.0         6.0         8.0        11.0         5.0   
13336         6.0        15.0         5.0         5.0         9.0        10.0   
...           ...         ...         ...         ...         ...         ...   
84730         0.0         NaN         NaN         0.0         0.0         NaN   
84731         0.0         0.0         0.0         0.0         0.0         0.0   
84732         0.0         0.0         0.0         0.0         0.0         0.0   
84733         0.0         0.0         0.0         0.0         NaN         0.0   
84734         0.0         0.0         0.0         0.0         1.0         NaN   

       2015-07-08  2015-07-09  ...  2016-12-23  2016-12-24  2016-12-25  \
13332        43.0        40.0  ...        74.0        34.0        62.0   
13333         6.0         6.0  ...        43.0        44.0        38.0   
13334         6.0        13.0  ...        44.0        47.0        46.0   
13335         5.0         6.0  ...        18.0        13.0        11.0   
13336         3.0         9.0  ...        32.0        40.0        38.0   
...           ...         ...  ...         ...         ...         ...   
84730         NaN         0.0  ...         7.0         8.0        15.0   
84731         0.0         0.0  ...        37.0         4.0        18.0   
84732         0.0         0.0  ...        48.0        13.0        24.0   
84733         0.0         0.0  ...        40.0         3.0        53.0   
84734         NaN         0.0  ...        26.0         1.0        39.0   

       2016-12-26  2016-12-27  2016-12-28  2016-12-29  2016-12-30  2016-12-31  \
13332        88.0        57.0        74.0        61.0        51.0        60.0   
13333        31.0        29.0        37.0        32.0        29.0        34.0   
13334        35.0        31.0        31.0        34.0        28.0        27.0   
13335         8.0         8.0        13.0         5.0        11.0         6.0   
13336        29.0        19.0        30.0        29.0        26.0        26.0   
...           ...         ...         ...         ...         ...         ...   
84730         7.0        18.0        17.0        15.0         3.0        11.0   
84731         2.0       172.0        19.0        29.0         2.0       178.0   
84732        11.0        12.0        51.0        23.0         3.0        48.0   
84733       300.0         3.0        44.0         6.0         0.0       154.0   
84734       346.0         2.0        28.0         4.0       163.0         1.0   

       Language  
13332      None  
13333      None  
13334      None  
13335      None  
13336      None  
...         ...  
84730      None  
84731      None  
84732      None  
84733      None  
84734      None  

[17855 rows x 552 columns]

In [54]:
train.head()

Page  2015-07-01  2015-07-02  \
0            2NE1_zh.wikipedia.org_all-access_spider        18.0        11.0   
1             2PM_zh.wikipedia.org_all-access_spider        11.0        14.0   
2              3C_zh.wikipedia.org_all-access_spider         1.0         0.0   
3         4minute_zh.wikipedia.org_all-access_spider        35.0        13.0   
4  52_Hz_I_Love_You_zh.wikipedia.org_all-access_s...         NaN         NaN   

   2015-07-03  2015-07-04  2015-07-05  2015-07-06  2015-07-07  2015-07-08  \
0         5.0        13.0        14.0         9.0         9.0        22.0   
1        15.0        18.0        11.0        13.0        22.0        11.0   
2         1.0         1.0         0.0         4.0         0.0         3.0   
3        10.0        94.0         4.0        26.0        14.0         9.0   
4         NaN         NaN         NaN         NaN         NaN         NaN   

   2015-07-09  ...  2016-12-23  2016-12-24  2016-12-25  2016-12-26  \
0        26.0  ...        63.0        15.0        26.0        14.0   
1        10.0  ...        42.0        28.0        15.0         9.0   
2         4.0  ...         1.0         1.0         7.0         4.0   
3        11.0  ...        10.0        26.0        27.0        16.0   
4         NaN  ...         9.0        25.0        13.0         3.0   

   2016-12-27  2016-12-28  2016-12-29  2016-12-30  2016-12-31  Language  
0        20.0        22.0        19.0        18.0        20.0        zh  
1        30.0        52.0        45.0        26.0        20.0        zh  
2         4.0         6.0         3.0         4.0        17.0        zh  
3        11.0        17.0        19.0        10.0        11.0        zh  
4        11.0        27.0        13.0        36.0        10.0        zh  

[5 rows x 552 columns]

1. Вікіпедії яких країн доступні в цьому датасеті?

In [55]:
list(filter(lambda item: item is not None, train["Language"].unique()))

['zh', 'fr', 'en', 'ru', 'de', 'ja', 'es']

Відповідь до завдання 1: **['zh', 'fr', 'en', 'ru', 'de', 'ja', 'es']**

*Думаю в завданні малось на увазі, на яких **мовах** доступні вікіпедії


2. Скільки сторінок з французької Вікіпедії у датасеті?

In [56]:
len(train[(train["Language"] == "fr")])

17802

Відповідь до завдання 2: **17802**

3. Вивести топ-3 статі для кожної країни за квітень 2016 року.


In [101]:
train_april_2016 = train.loc[:, (train.columns >= '2016-04-01') & (train.columns <= '2016-04-30')]
train_april_2016

2016-04-01  2016-04-02  2016-04-03  2016-04-04  2016-04-05  \
0             12.0        14.0        13.0         9.0       490.0   
1             25.0        18.0        18.0        16.0        20.0   
2              4.0       155.0       155.0        83.0        48.0   
3             18.0        18.0        23.0        27.0        39.0   
4              NaN         NaN         NaN         NaN         NaN   
...            ...         ...         ...         ...         ...   
145058         NaN         NaN         NaN         NaN         NaN   
145059         NaN         NaN         NaN         NaN         NaN   
145060         NaN         NaN         NaN         NaN         NaN   
145061         NaN         NaN         NaN         NaN         NaN   
145062         NaN         NaN         NaN         NaN         NaN   

        2016-04-06  2016-04-07  2016-04-08  2016-04-09  2016-04-10  ...  \
0            189.0       102.0        38.0       126.0        71.0  ...   
1             17.0        16.0        13.0        15.0        19.0  ...   
2             31.0        16.0         6.0        13.0         8.0  ...   
3             11.0        16.0         9.0        26.0        14.0  ...   
4              NaN         NaN         NaN         NaN         NaN  ...   
...            ...         ...         ...         ...         ...  ...   
145058         NaN         NaN         NaN         NaN         NaN  ...   
145059         NaN         NaN         NaN         NaN         NaN  ...   
145060         NaN         NaN         NaN         NaN         NaN  ...   
145061         NaN         NaN         NaN         NaN         NaN  ...   
145062         NaN         NaN         NaN         NaN         NaN  ...   

        2016-04-21  2016-04-22  2016-04-23  2016-04-24  2016-04-25  \
0             44.0         6.0        31.0        17.0        25.0   
1             39.0        17.0        17.0        41.0        35.0   
2              8.0         4.0         6.0         5.0         2.0   
3             12.0        14.0        14.0        23.0        21.0   
4              1.0        10.0         9.0         2.0         0.0   
...            ...         ...         ...         ...         ...   
145058         NaN         NaN         NaN         NaN         NaN   
145059         NaN         NaN         NaN         NaN         NaN   
145060         NaN         NaN         NaN         NaN         NaN   
145061         NaN         NaN         NaN         NaN         NaN   
145062         NaN         NaN         NaN         NaN         NaN   

        2016-04-26  2016-04-27  2016-04-28  2016-04-29  2016-04-30  
0             40.0        19.0        15.0        15.0        29.0  
1             16.0         9.0        64.0        22.0        22.0  
2              7.0         3.0         7.0         6.0         3.0  
3             11.0        19.0         9.0        10.0        11.0  
4              5.0         0.0         3.0        55.0       234.0  
...            ...         ...         ...         ...         ...  
145058         NaN         NaN         NaN         NaN         NaN  
145059         NaN         NaN         NaN         NaN         NaN  
145060         NaN         NaN         NaN         NaN         NaN  
145061         NaN         NaN         NaN         NaN         NaN  
145062         NaN         NaN         NaN         NaN         NaN  

[145063 rows x 30 columns]

In [115]:
train_april_2016['Page'] = train['Page']
train_april_2016['Language'] = train['Language']

In [117]:
train_april_2016 = train_april_2016.sort_values('Total', ascending=False)
train_april_2016

2016-04-01  2016-04-02  2016-04-03  2016-04-04  2016-04-05  \
38573   13658940.0  15130113.0  16171487.0  18152644.0  17002940.0   
9774     8091205.0   9488202.0  10427940.0  12531466.0  11470276.0   
74114    5383386.0   5443574.0   5535088.0   5432117.0   5351041.0   
139119   3026671.0   2885997.0   2897868.0   3112302.0   3097582.0   
39180    2087270.0   2016520.0   1829010.0   2089875.0   2141175.0   
...            ...         ...         ...         ...         ...   
32116          NaN         NaN         NaN         NaN         NaN   
32115          NaN         NaN         NaN         NaN         NaN   
32114          NaN         NaN         NaN         NaN         NaN   
32112          NaN         NaN         NaN         NaN         NaN   
145062         NaN         NaN         NaN         NaN         NaN   

        2016-04-06  2016-04-07  2016-04-08  2016-04-09  2016-04-10  ...  \
38573   17028051.0  16822207.0  16176598.0  17166429.0  17373626.0  ...   
9774    11547859.0  11358696.0  10714095.0  11646912.0  11770144.0  ...   
74114    5302913.0   5289973.0   5292426.0   5340459.0   5413446.0  ...   
139119   3070997.0   3061373.0   3010177.0   2777728.0   2925551.0  ...   
39180    2174616.0   2150831.0   2020588.0   2501728.0   1821356.0  ...   
...            ...         ...         ...         ...         ...  ...   
32116          NaN         NaN         NaN         NaN         NaN  ...   
32115          NaN         NaN         NaN         NaN         NaN  ...   
32114          NaN         NaN         NaN         NaN         NaN  ...   
32112          NaN         NaN         NaN         NaN         NaN  ...   
145062         NaN         NaN         NaN         NaN         NaN  ...   

        2016-04-24  2016-04-25  2016-04-26  2016-04-27  2016-04-28  \
38573   17780974.0  20181132.0  17808242.0  17623987.0  17918962.0   
9774    12283817.0  14716730.0  12438996.0  12233641.0  12473206.0   
74114    5353261.0   5339136.0   5246598.0   5270902.0   5326951.0   
139119   2852765.0   3008092.0   2992540.0   2948328.0   3045567.0   
39180    1837434.0   2126110.0   2397769.0   2193427.0   2343212.0   
...            ...         ...         ...         ...         ...   
32116          NaN         NaN         NaN         NaN         NaN   
32115          NaN         NaN         NaN         NaN         NaN   
32114          NaN         NaN         NaN         NaN         NaN   
32112          NaN         NaN         NaN         NaN         NaN   
145062         NaN         NaN         NaN         NaN         NaN   

        2016-04-29  2016-04-30         Total  \
38573   16022991.0  17187067.0  4.038741e+09   
9774    10640040.0  11777106.0  2.744305e+09   
74114    5263527.0   5285097.0  1.258599e+09   
139119   3014146.0   2620346.0  6.857226e+08   
39180    2033243.0   1805432.0  5.006866e+08   
...            ...         ...           ...   
32116          NaN         NaN  0.000000e+00   
32115          NaN         NaN  0.000000e+00   
32114          NaN         NaN  0.000000e+00   
32112          NaN         NaN  0.000000e+00   
145062         NaN         NaN  0.000000e+00   

                                                     Page  Language  
38573    Main_Page_en.wikipedia.org_all-access_all-agents        en  
9774        Main_Page_en.wikipedia.org_desktop_all-agents        en  
74114    Main_Page_en.wikipedia.org_mobile-web_all-agents        en  
139119  Wikipedia:Hauptseite_de.wikipedia.org_all-acce...        de  
39180   Special:Search_en.wikipedia.org_all-access_all...        en  
...                                                   ...       ...  
32116         星際過客_zh.wikipedia.org_all-access_all-agents        zh  
32115        下女的誘惑_zh.wikipedia.org_all-access_all-agents        zh  
32114   決戰異世界：弒血之戰_zh.wikipedia.org_all-access_all-agents        zh  
32112      多桑の純萃年代_zh.wikipedia.org_all-access_all-agents        zh  
145062  Francisco_el_matemático_(serie_de_televisión_d...        es  

[145063

In [132]:
# Group the DataFrame by 'Language'
grouped = train_april_2016.groupby('Language')

# Initialize an empty list to store the top pages for each language
top_pages = []

# Iterate over each group (language)
for language, group in grouped:
    # Sort the group by 'Total' column in descending order and get the top 3 pages
    top_3_pages = group.sort_values('Total', ascending=False).head(3)
    # Append the top pages to the 'top_pages' list
    top_pages.append(top_3_pages)

# Concatenate the top pages into a single DataFrame
result = pd.concat(top_pages)

# Display the top pages for each language
result[["Page", "Language","Total"]]

Page Language  \
139119  Wikipedia:Hauptseite_de.wikipedia.org_all-acce...       de   
116196  Wikipedia:Hauptseite_de.wikipedia.org_mobile-w...       de   
67049   Wikipedia:Hauptseite_de.wikipedia.org_desktop_...       de   
38573    Main_Page_en.wikipedia.org_all-access_all-agents       en   
9774        Main_Page_en.wikipedia.org_desktop_all-agents       en   
74114    Main_Page_en.wikipedia.org_mobile-web_all-agents       en   
92205   Wikipedia:Portada_es.wikipedia.org_all-access_...       es   
95855   Wikipedia:Portada_es.wikipedia.org_mobile-web_...       es   
90810   Especial:Buscar_es.wikipedia.org_all-access_al...       es   
27330   Wikipédia:Accueil_principal_fr.wikipedia.org_a...       fr   
26993   Organisme_de_placement_collectif_en_valeurs_mo...       fr   
7213    Organisme_de_placement_collectif_en_valeurs_mo...       fr   
120336      メインページ_ja.wikipedia.org_all-access_all-agents       ja   
86431          メインページ_ja.wikipedia.org_desktop_all-agents       ja   
57309       メインページ_ja.wikipedia.org_mobile-web_all-agents       ja   
99322   Заглавная_страница_ru.wikipedia.org_all-access...       ru   
103123  Заглавная_страница_ru.wikipedia.org_desktop_al...       ru   
17670   Заглавная_страница_ru.wikipedia.org_mobile-web...       ru   
28727   Wikipedia:首页_zh.wikipedia.org_all-access_all-a...       zh   
61350    Wikipedia:首页_zh.wikipedia.org_desktop_all-agents       zh   
28728   Special:搜索_zh.wikipedia.org_all-access_all-agents       zh   

               Total  
139119  6.857226e+08  
116196  4.974432e+08  
67049   1.697752e+08  
38573   4.038741e+09  
9774    2.744305e+09  
74114   1.258599e+09  
92205   3.217345e+08  
95855   2.461088e+08  
90810   1.027164e+08  
27330   3.741841e+08  
26993   2.925371e+08  
7213    2.925153e+08  
120336  8.423367e+07  
86431   5.211122e+07  
57309   2.918886e+07  
99322   3.004519e+08  
103123  1.548645e+08  
17670   1.406090e+08  
28727   5.190821e+07  
61350   2.972896e+07  
28728   2.226140e+07

4. Знайти сторінку англійської вікіпедії з найбільшим середнім/медіаною/дисперсією за липень 2016 року.
5. Які статі отримали найбільший буст по переглядам за травень 2016 року в абсолютних показниках і відносних. (розглядати сторінки щонайменше з 1000 переглядами)
6. Візуалізувати кількість переглядів для топ-5 сторінок англійської Вікіпедії на одному графіку за перше півріччя 2016 року.
7. Порівняти розподіли німецьких і французьких сторінок вікіпедії за червень 2016 року.
8. Порівняти кількість переглядів для сторінок японської вікіпедії з розбивкою по дням тижня.